src/crypto/crypto_util.h

```diff
@@ -0,0 +1,680 @@
+         arg->IsArrayBuffer() ||
+         arg->IsSharedArrayBuffer();
+}
+
+template <typename T>
+class ArrayBufferOrViewContents {
+ public:
+  ArrayBufferOrViewContents() = default;
+
+  inline explicit ArrayBufferOrViewContents(v8::Local<v8::Value> buf) {
+    CHECK(IsAnyByteSource(buf));
+    if (buf->IsArrayBufferView()) {
+      auto view = buf.As<v8::ArrayBufferView>();
+      offset_ = view->ByteOffset();
+      length_ = view->ByteLength();
+      store_ = view->Buffer()->GetBackingStore();
+    } else if (buf->IsArrayBuffer()) {
+      auto ab = buf.As<v8::ArrayBuffer>();
+      offset_ = 0;
+      length_ = ab->ByteLength();
+      store_ = ab->GetBackingStore();
+    } else {
+      auto sab = buf.As<v8::SharedArrayBuffer>();
+      offset_ = 0;
+      length_ = sab->ByteLength();
+      store_ = sab->GetBackingStore();
+    }
+  }
+
+  inline const T* data() const {
+    // Ideally, these would return nullptr if IsEmpty() or length_ is zero,
+    // but some of the openssl API react badly if given a nullptr even when
+    // length is zero, so we have to return something.
+    if (size() == 0)
+      return &buf;
+    return reinterpret_cast<T*>(store_->Data()) + offset_;
+  }
+
+  inline T* data() {
+    // Ideally, these would return nullptr if IsEmpty() or length_ is zero,
+    // but some of the openssl API react badly if given a nullptr even when
+    // length is zero, so we have to return something.
+    if (size() == 0)
+      return &buf;
+    return reinterpret_cast<T*>(store_->Data()) + offset_;
+  }
+
+  inline size_t size() const { return length_; }
+
+  // In most cases, input buffer sizes passed in to openssl need to
+  // be limited to <= INT_MAX. This utility method helps us check.
+  inline bool CheckSizeInt32() { return size() <= INT_MAX; }
+
+  inline ByteSource ToByteSource() const {
+    return ByteSource::Foreign(data(), size());
+  }
+
+  inline ByteSource ToCopy() const {
+    if (size() == 0) return ByteSource();
+    char* buf = MallocOpenSSL<char>(size());
+    CHECK_NOT_NULL(buf);
+    memcpy(buf, data(), size());
+    return ByteSource::Allocated(buf, size());
+  }
+
+  inline ByteSource ToNullTerminatedCopy() const {
+    if (size() == 0) return ByteSource();
+    char* buf = MallocOpenSSL<char>(size() + 1);
+    CHECK_NOT_NULL(buf);
+    buf[size()] = 0;
+    memcpy(buf, data(), size());
+    return ByteSource::Allocated(buf, size());
+  }
+
+  template <typename M>
+  void CopyTo(std::vector<M>* vec) const { CopyBuffer(this, vec); }
+
+  template <typename M>
+  void CopyTo(M* dest, size_t len) const {
+    len = std::min(len, size());
+    if (len > 0 && data() != nullptr)
+      memcpy(dest, data(), len);
+  }
+
+ private:
+  T buf = 0;
+  size_t offset_ = 0;
+  size_t length_ = 0;
+  std::shared_ptr<v8::BackingStore> store_;
+};
+
+template <typename T>
+void CopyBuffer(
+    const ArrayBufferOrViewContents<T>& buf,
+    std::vector<T>* vec) {
+  vec->clear();
+  vec->resize(buf.size());
```

Comment:
This is a bit dangerous if `sizeof(M) != 1`, maybe add a `static_assert`? Also, both `CopyBuffer` variants could just return the `std::vector<>` directly, that’s possibly more efficient but also generally just a bit more straightforward